### PR TITLE
worker: enable distributed tracing for dbworker handler jobs

### PR DESCRIPTION
**Pull request by Strum355**
## What it do

This PR enables `worker`/`dbworker`s to emit traces to Jaeger. A span context is created whenever a job is successfully read from the database, and then it's context propagation as normal.

## Sampling of traces

In its initial stage, at least, this PR does not have an answer on how to handle sampling of the vast increase of traces that would be generated. Currently, its effectively got a sampling strategy of `"all"`, which would probably be unwise in production. Initial discussions with @efritz considered [Jaeger 2.20's "delayed sampling"](https://pkg.go.dev/github.com/uber/jaeger-client-go#readme-delayed-sampling) capabilities, but its unclear to us whether our setup/customer setups are setup for this, as well as concerns with its inter-process capabilities.

If these turn out to be non-issues or solvable issues, workers would/should create a custom non-global tracer that uses this delayed sampling instead of [tracer.GlobalTracer()](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24%40352f5d6+Tracer:+%26trace.Tracer%7BTracer:+opentracing.GlobalTracer%28%29%7D%2C&patternType=regexp) that the rest of the service (aka non-worker codepaths) would be using. This non-global tracer would be passed to [workerutil.NewMetrics()](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@nsc/worker-tracing-inject/-/blob/internal/workerutil/observability.go?L31:6), as ultimately consumed by either [dbworker.NewWorker()](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@352f5d653aa304f86f8e3062cacde7e316c3d50b/-/blob/internal/workerutil/dbworker/worker.go?L10:6) or [workerutil.NewWorker()](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@nsc/worker-tracing-inject/-/blob/internal/workerutil/worker.go?L78:6). Some parts of the tracing code (probably [internal/trace/ot](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@nsc/worker-tracing-inject/-/blob/internal/trace/ot/ot.go), have not done a proper analysis of what areas) may need to be revamped to allow the tracer used to create a new child span come from any existing spans, using [opentracing.Span.Tracer()](https://pkg.go.dev/github.com/uber/jaeger-client-go#Span.Tracer).

### CODE REVIEW

File: internal/workerutil/worker.go [hunk 1]
```diff
@@ -261,17 +263,21 @@ func (w *Worker) dequeueAndHandle() (dequeued bool, err error) {
 	log15.Debug("Dequeued record for processing", "name", w.options.Name, "id", record.RecordID())
 
 	if hook, ok := w.handler.(WithHooks); ok {
-		hook.PreHandle(handleCtx, record)
+		preCtx, endObservation := w.options.Metrics.operations.preHandle.With(workerCtxWithSpan, nil, observation.Args{})
```
Comment (eseliger) [inline internal/workerutil/worker.go:266]: I think this should still use handleCtx, as the `cancel` function will be invoked for that on job cancel, not for `w.ctx`.
```suggestion
		preCtx, endObservation := w.options.Metrics.operations.preHandle.With(handleCtx, nil, observation.Args{})
```

File: internal/workerutil/worker.go [hunk 2]
```diff
@@ -261,17 +263,21 @@ func (w *Worker) dequeueAndHandle() (dequeued bool, err error) {
...
 
 	w.wg.Add(1)
 
 	go func() {
 		defer func() {
 			if hook, ok := w.handler.(WithHooks); ok {
+				postCtx, endObservation := w.options.Metrics.operations.postHandle.With(workerCtxWithSpan, nil, observation.Args{})
+				defer endObservation(1, observation.Args{})
 				// Don't use handleCtx here, the record is already not owned by
 				// this worker anymore at this point.
```
Comment (eseliger) [inline internal/workerutil/worker.go:279]: Consider moving that comment to above the postHandle.With

File: internal/workerutil/worker.go [hunk 3]
```diff
@@ -261,17 +263,21 @@ func (w *Worker) dequeueAndHandle() (dequeued bool, err error) {
...
 	}
 
 	w.wg.Add(1)
 
 	go func() {
 		defer func() {
 			if hook, ok := w.handler.(WithHooks); ok {
+				postCtx, endObservation := w.options.Metrics.operations.postHandle.With(workerCtxWithSpan, nil, observation.Args{})
+				defer endObservation(1, observation.Args{})
 				// Don't use handleCtx here, the record is already not owned by
```
Comment (eseliger) [inline internal/workerutil/worker.go:278]:
```suggestion
				// Don't use workerCtxWithSpan here, the record is already not owned by
```

**Review by eseliger — APPROVED** [review]
🚀 🚀 🚀

**Review by tsenart — APPROVED** [review]
It'd be great to log this trace id so we can jump to these traces from logs. Check how we do it with the HTTPTraceMiddleware.

**Comment by bobheadxi** [comment]
> considered Jaeger 2.20's "delayed sampling" capabilities, but its unclear to us whether our setup/customer setups are setup for this, as well as concerns with its inter-process capabilities.

We seem to be using jaeger 1.24 (https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/docker-images/jaeger-agent/build.sh?L7) so that might be problematic 🤔

**Comment by bobheadxi** [comment]
> considered Jaeger 2.20's "delayed sampling" capabilities, but its unclear to us whether our setup/customer setups are setup for this, as well as concerns with its inter-process capabilities.

What is needed to get this set up? The linked docs don't seem to mention anything that needs to happen serverside

**Comment by Strum355** [comment]
> > considered Jaeger 2.20's "delayed sampling" capabilities, but its unclear to us whether our setup/customer setups are setup for this, as well as concerns with its inter-process capabilities.
> 
> What is needed to get this set up? The linked docs don't seem to mention anything that needs to happen serverside

Youre right, turns out its a client side feature. What seems to be important to know here is the following excerpt from the jaeger tracking issue:
> This approach does not allow sampling caller service spans based on a sampling decision made in the downstream service, because that information is not propagated backwards (although it the errors are propagated and set as tags in the caller service then it may still affect sampling).

If Im reading this right, some care would have to be taken if crossing process boundaries that we either 1) start a new **trace** when crossing process boundaries (most applicable if the operation is async to the source) or 2) an error is propagated back up so that the calling service can make the same delayed sampling decision to sample a trace. In the case of no error, a predetermined sampling possibility as is already standard should be fine, something like per-operation which is a pre-existing sampling strategy perhaps or we can keep it simple for now @bobheadxi what do you think?

**Review by bobheadxi — APPROVED** [review]

**Pull request merged**